### PR TITLE
snap-bootstrap: expand data partition on install

### PR DESCRIPTION
**Pull request by cmatsuoka**
Expand the partition with role system-data if it's the last partition
and there's empty disk space available.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

### CODE REVIEW

File: cmd/snap-bootstrap/partition/sfdisk.go [hunk 1]
```diff
@@ -235,13 +233,23 @@ func deviceName(name string, index int) string {
...
+func buildPartitionList(dl *DeviceLayout, pv *gadget.LaidOutVolume) (sfdiskInput *bytes.Buffer, toBeCreated []DeviceStructure) {
+	ptable := dl.partitionTable
 	// Keep track what partitions we already have on disk
 	seen := map[uint64]bool{}
 	for _, p := range ptable.Partitions {
 		seen[p.Start] = true
 	}
 
+	// Check if the last partition has a system-data role
+	expandData := false
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/partition/sfdisk.go]: `s/expandData/canExpandData/`

File: cmd/snap-bootstrap/partition/sfdisk.go [hunk 2]
```diff
@@ -255,12 +263,19 @@ func buildPartitionList(ptable *sfdiskPartitionTable, pv *gadget.LaidOutVolume)
 		if seen[uint64(start)] {
 			continue
 		}
+
+		// Check if the data partition should be expanded
+		size := s.Size
+		if s.Role == gadget.SystemData && expandData && p.StartOffset+s.Size < dl.Size {
+			size = dl.Size - p.StartOffset
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/partition/sfdisk.go]: IIRC `last-lba` is the last usable sector, so we could`+ dl.SectorSize` here?. OTOH, it's only a single sector, so there's negligible win here. Also, it'd be worth to check whether sfdisk complains about it first.

File: cmd/snap-bootstrap/partition/sfdisk_test.go [hunk 3]
```diff
@@ -263,9 +263,13 @@ func (s *partitionTestSuite) TestBuildPartitionList(c *C) {
 	pv, err := gadget.PositionedVolumeFromGadget(gadgetRoot)
 	c.Assert(err, IsNil)
 
-	sfdiskInput, created := partition.BuildPartitionList(ptable, pv)
+	dl, err := partition.DeviceLayoutFromPartitionTable(ptable)
+	c.Assert(err, IsNil)
+
+	// the expected expanded writable partition size is (1200M + 8388574*512 - (2M + 1200M + 1200M))/512
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/partition/sfdisk_test.go]: Maybe tweak the formula to match the check in `if {}` ? Something like:
```
// the expected expanded writable partition size is calculated like this:
// start offset = (2M + 1200M), size = 1200M, expanded size = (8388574*512 - (start + size))
```

File: cmd/snap-bootstrap/partition/sfdisk.go (hunk 2 above)
Comment (cmatsuoka) [reply cmd/snap-bootstrap/partition/sfdisk.go]: In this case I believe DeviceLayout.Size should be fixed to use `(ptable.LastLBA + 1) * sectorSize` and the size computation here should remain the same. Will fix it there.

**Comment by mvo5** [comment]
A bit of a meta-comment. I think it would be nice if we would be explicit here (best to double check with Gustavo on that too though). I.e. right now we define ubuntu-data as:
```
 - name: ubuntu-data
        role: system-data
        filesystem: ext4
        type: 83,0FC63DAF-8483-4772-8E79-3D69D8477DE4
        # XXX: make auto-grow to partition
        size: 3G
```
It feels to me something like `auto-grow: true` or `size: 3G+` would be good to have(syntax needs Gustavo input I think). WDYT?

File: cmd/snap-bootstrap/bootstrap/bootstrap.go [hunk 4]
```diff
@@ -149,7 +149,14 @@ func ensureLayoutCompatibility(gadgetLayout *gadget.LaidOutVolume, diskLayout *p
 	eq := func(ds partition.DeviceStructure, gs gadget.LaidOutStructure) bool {
 		dv := ds.VolumeStructure
 		gv := gs.VolumeStructure
-		return dv.Name == gv.Name && ds.StartOffset == gs.StartOffset && dv.Size == gv.Size && dv.Filesystem == gv.Filesystem
+		check := dv.Name == gv.Name && ds.StartOffset == gs.StartOffset && dv.Filesystem == gv.Filesystem
+
+		// Don't test system-data size, it could have been expanded
+		if gv.Role == gadget.SystemData {
+			return check
```
Comment (bboozzoo) [inline cmd/snap-bootstrap/bootstrap/bootstrap.go]: Should this be `check && dv.Size >= gv.Size` then?

Comment (cmatsuoka) [reply cmd/snap-bootstrap/bootstrap/bootstrap.go]: It makes sense to enforce the minimum size of the partition.

**Comment by cmatsuoka** [comment]
> It feels to me something like `auto-grow: true` or `size: 3G+` would be good to have(syntax needs Gustavo input I think). WDYT?

I like the idea of being explicit instead of hard-coding, however note that there are many other things already hard-coded (labels for certain roles, which partitions can be created at install time, which partitions can be encrypted). At this point, I'm not sure that it would be worth the extra risk/effort.

**Comment by cmatsuoka** [comment]
One advantage of making it explicit would be to have the possibility to choose if we want it to grow or not.

**Comment by cmatsuoka** [comment]
Option suggested by Michael: explicitly state when we _don't_ want it to be expanded.

**Comment by mvo5** [comment]
About my auto-grow suggestion: my original thinking was that implicit special behavior like auto-grow for ubuntu-data would be a bit magic. But then Claudio said that we already do a lot of implicit things to the ubuntu-data partition so it is already pretty special. This makes sense so adding this extra gadget.yaml syntax now does not make that much sense I think. If we find a use-case for ubuntu-data to *not* auto-grow  we can always add a new "auto-grow: false" yaml.

**Pull request merged**